### ISSUE TITLE
Are the ids generated guaranteed to be unique?

### ISSUE THREAD

**Issue by mithun-daa** [substack/node-hat#8]
Are the IDs generated using:
```
var hat = require('hat');

var id = hat();
console.log(id);
```
guaranteed to be unique?
Also, is hat a good solution for creating API keys?

**Comment by bendrucker** [comment]
No, `hat` is not a singleton. You have to create a rack for uniqueness. And yes, you'll generally be okay using these as API keys.

**Comment by markstos** [comment]
This issue can be closed.

**Comment by ranjithav-tzl** [comment]
var hat = require('hat');
var rack = hat.rack();

console.log(rack());
console.log(rack());


Maximum how many unique rack we can create using this code?

**Comment by nolanamy** [comment]
**TL;DR: No.**

They're only guaranteed to be unique if you're using a `rack` (which simply keeps track of which ids you've used so far) and _the same rack for every id ever generated_. For a web app, this requires somehow persisting and sharing the rack between app instances. I have no idea how node-hat proposes you do that ¯\\\_(ツ)\_/¯.

If you want to guarantee that your ids are unique (for session tokens or API keys, for example), either use a cryptographically secure generator (bonus! ids will be much harder to guess), or check each new id against all your existing ids (stored in your database, for example). So... use a cryptographically secure generator.

We used `hat()` to generate our first 600,000 or so session tokens. A few users complained about seeing someone else's stuff... not good. Turned out there were 202 duplicated tokens in the 623,301 generated with `hat()`. Bad. With 128 bits, the probability that the 623,301st id is a duplicate _should be_ 1 in 5.46 x 10^32, and you should expect to see your first duplicate after generating about 2^64 (or 18 quintillion) ids.

So we switched to `crypto.randomBytes`:

```javascript
crypto.randomBytes(16, function (err, bytes) {
  ...
  var token = bytes.toString('hex');
  ...
});
```
https://nodejs.org/api/crypto.html#crypto_crypto_randombytes_size_callback

...just checked – about a million and a half sessions total, no more duplicates. yay.

**Comment by benfjohnson** [comment]
@nolanamy thanks for the alternative!

**Comment by BDiehr** [comment]
So it sounds like people should just use https://github.com/broofa/node-uuid instead?

**Comment by lokesh010** [comment]
We can concat Date with 

> **TL;DR: No.**
> 
> They're only guaranteed to be unique if you're using a `rack` (which simply keeps track of which ids you've used so far) and _the same rack for every id ever generated_. For a web app, this requires somehow persisting and sharing the rack between app instances. I have no idea how node-hat proposes you do that ¯_(ツ)_/¯.
> 
> If you want to guarantee that your ids are unique (for session tokens or API keys, for example), either use a cryptographically secure generator (bonus! ids will be much harder to guess), or check each new id against all your existing ids (stored in your database, for example). So... use a cryptographically secure generator.
> 
> We used `hat()` to generate our first 600,000 or so session tokens. A few users complained about seeing someone else's stuff... not good. Turned out there were 202 duplicated tokens in the 623,301 generated with `hat()`. Bad. With 128 bits, the probability that the 623,301st id is a duplicate _should be_ 1 in 5.46 x 10^32, and you should expect to see your first duplicate after generating about 2^64 (or 18 quintillion) ids.
> 
> So we switched to `crypto.randomBytes`:
> 
> ```js
> crypto.randomBytes(16, function (err, bytes) {
>   ...
>   var token = bytes.toString('hex');
>   ...
> });
> ```
> 
> https://nodejs.org/api/crypto.html#crypto_crypto_randombytes_size_callback
> 
> ...just checked – about a million and a half sessions total, no more duplicates. yay.

We can just concat date with the byte if you care about duplicacy even though chances are very little.

**Comment by lokesh010** [comment]
We can just concat date with the byte if you care about duplicacy even though chances are very little.
